### PR TITLE
fix(api-keys): restore project base/legacy API key rotation (#4891)

### LINKED ISSUE #4891: Restore project base/legacy API key rotation (dropped in the unified-keys rework #3386)
## For Humans

A project base (default/legacy) API key can no longer be rotated from the product UI. The unified-API-keys rework (PR 3386) made the legacy project key read-only and dropped the rotation entry point (the old dialog was orphaned). The reporting was right that the capability is gone; the nuance is it was removed deliberately, not by a runtime bug.

**Decision (Drew): base-key rotation MUST still exist.** A leaked or compromised base key has to be rotatable in-product. So this is a capability regression to fix: restore a rotate path for the legacy project key.

**Ask:** restore a supported way to rotate a project base/legacy API key in the UI (wire the existing, still-working regenerate mutation into the current unified-keys page, or a restored entry point), permission-gated and audited.

_Status: bug, P1 — fix DECIDED. Investigation complete; ACs below. Plan/spec + implementation next. not-ready until the spec step lands._

<details><summary>For Agents — investigation + ACs</summary>

## Investigation

**Root cause (origin/main, verified via gh + the unified-keys spec):** NOT a broken mutation — the base-key rotation UI was intentionally removed in the unified-API-keys rework (PR 3386).

- RegenerateApiKeyDialog (src/components/settings/) and the useRegenerateApiKey hook have ZERO importers — orphaned dead code (confirmed: not rendered in the UI).
- The regenerateApiKey tRPC mutation (project.ts:255) still EXISTS and is not broken — it is simply never called, because nothing renders its UI. It is permission-gated (project:manage), generates a key, writes project.apiKey via Prisma, and audit-logs.
- The unified-api-keys spec (specs/api-keys/unified-api-keys.feature) treats the project base key as a LEGACY key: shown in the table with NO edit or revoke button (line 78), kept working for backward-compat auth, never deleted by new key ops. Read-only by design.
- project.apiKey is still read by live auth paths (auth-cli, misc routes, scenarios orchestrator, simulation runner), so the column is live — the only thing missing is a UI path to rotate it.

**Dropped (wrong-layer) hypotheses:** permission-gate / key-format / audit-log mutation failures — all moot, the mutation is never reached. Recorded so the next reader does not re-chase them.

**Fix direction (for the plan step, not prescriptive):** the still-working regenerate mutation needs a UI entry point on the current unified-keys surface; the orphaned dialog/hook can be revived or reimplemented. The plan step decides which.

## Acceptance Criteria

1. **Rotation available in the UI** — a user with project:manage permission can rotate the project base/legacy API key from the product UI (a control on the unified-keys page or a restored entry point). Evidence: an integration/Playwright test (or attached screenshot) showing the control and a successful rotation issuing a new key.
2. **Rotation takes effect** — after rotation, the previous base key fails authentication and the new key authenticates. Evidence: integration test — a request with the old key returns 401/403; the same request with the new key returns 200.
3. **Permission-gated** — a user WITHOUT project:manage cannot rotate (control hidden/disabled AND the mutation rejects with 403). Evidence: a test asserting the 403 for a non-manage user.
4. **Legacy backward-compat preserved (regression surface)** — the legacy project key continues to authenticate until it is rotated; the existing unified-api-keys.feature legacy-project-key scenarios remain green. Evidence: those scenarios pass.
5. **Audited** — rotation writes an audit-log entry (project.apiKey.regenerated), preserving the current mutation behavior. Evidence: a test asserting the audit-log entry on rotation.
6. **Failure-handling (no lockout)** — if the rotation write fails, the existing key stays valid and the user sees an error (no half-rotated or locked-out state). Evidence: a test simulating a write failure where the old key still authenticates and an error is surfaced.

</details>

### PR DESCRIPTION
**Pull request by drewdrewthis**
## What & why

Fixes #4891 (P1 security / capability regression).

The unified-API-keys rework (#3386) intentionally made the legacy **project base key** read-only and dropped its rotation entry point, orphaning `RegenerateApiKeyDialog` + the `useRegenerateApiKey` hook (zero importers). The `project.regenerateApiKey` tRPC mutation kept working — `project:manage`-gated and audited — but was **unreachable** from the UI. Result: a leaked/compromised project base key could not be rotated in-product.

**Decision (Drew):** base-key rotation must still exist. This wires the existing, still-working mutation back into the current unified API Keys page.

## Approach

- Added a `project:manage`-gated **Rotate** action (`RotateCw`) to the legacy "Project API Key" row in `ApiKeysSection.tsx`. The row keeps its no-edit / no-revoke shape — rotation is the only mutating affordance, and only when permitted.
- **Reused** the orphaned `RegenerateApiKeyDialog` for the confirm step (revived, not duplicated).
- Wired `api.project.regenerateApiKey.useMutation` **inline**, mirroring the page's sibling `create`/`update`/`revoke` handlers (toast + `organization.getAll` invalidation). The rotated key is surfaced once via the existing `TokenCreatedDialog`.
- **Deleted** the now-redundant `useRegenerateApiKey` hook — its logic is inlined to match the file's other handlers; keeping both would be two ways to do one thing.
- The `project.regenerateApiKey` mutation itself is **unchanged** (single atomic `prisma.project.update` → audit log).

## Acceptance criteria → evidence

| AC | Evidence | Result |
|----|----------|--------|
| **1. Rotation available in UI** | `project-key-rotation.integration.test.tsx`: control present on legacy row, opens confirm dialog, confirming calls mutation with `{ projectId }` | ✅ component 6/6 |
| **2. Rotation takes effect** | `project.regenerateApiKey.integration.test.ts`: after rotation the old raw key no longer resolves via the real `TokenResolver` legacy path; the new key resolves scoped to the project | ✅ integration 5/5 |
| **3. Permission-gated** | component: no rotate control when `!hasPermission('project:manage')`; integration: viewer (org MEMBER + team VIEWER) → `UNAUTHORIZED`, base key unchanged | ✅ |
| **4. Legacy back-compat preserved** | component: legacy row still has no edit/revoke control; `token-resolver.unit.test.ts` green; auth code untouched | ✅ token-resolver 3/3 |
| **5. Audited** | `…unit.test.ts`: `auditLog({ action: "project.apiKey.regenerated" })` asserted; `…integration.test.ts`: audit row exists after rotation | ✅ unit 6/6 |
| **6. Failure-handling (no lockout)** | `…unit.test.ts`: failed `update` → no `project.apiKey.regenerated` audit (single atomic write leaves the row, so the old key stays valid); component: error toast surfaced | ✅ |

## Local verification

```
typecheck (tsgo)                          clean (exit 0)
project.regenerateApiKey.unit.test.ts     6 passed
project.regenerateApiKey.integration.ts   5 passed   (real Postgres)
project-key-rotation.integration.test.tsx 6 passed
api-keys __tests__ (full dir)            37 passed   (incl. existing ingestion-split + scope-filter)
token-resolver.unit.test.ts               3 passed
```

## Falsifiability proof (red → green)

Both new test files are **load-bearing** — they fail against the pre-fix tree and pass with the fix. Since this PR is on hold (no push), the "tests-only commit" half is reproduced **locally**: this PR's tests at HEAD (`8a74f831b`) are run against the two production files reverted to the merge-base with `main` (`1f8a32a4a`) — exactly what a tests-only commit stacked on `main` would exercise — then the files are restored (no code changed, nothing pushed). Counts are `7/7` each at HEAD; each suite grew by one test during review-thread hardening, so they read one higher than the `6/6` snapshots above.

| New test file | Pre-fix (prod @ `1f8a32a4a`, tests @ HEAD) | With fix (HEAD `8a74f831b`) | CI on HEAD |
|---|---|---|---|
| `project-key-rotation.integration.test.tsx` — UI | 🔴 5 failed / 2 passed | 🟢 7 passed | 🟢 `test-integration` |
| `project.regenerateApiKey.unit.test.ts` — server | 🔴 1 failed / 6 passed | 🟢 7 passed | 🟢 `test-unit` |

**Revert `ApiKeysSection.tsx`** (drops the `project:manage`-gated Rotate control) → the 5 behavior tests go red, while the 2 *absence* assertions stay green — so the suite isn't trivially all-or-nothing:

```
× offers a control to rotate the project base API key
× opens the regenerate confirmation dialog
× calls the regenerate mutation with the project id
× reveals the new key in the Token Created dialog
× shows an error toast explaining the rotation did not happen
✓ does not offer a control to rotate … (no project:manage)
✓ does not offer edit or revoke controls on the legacy row
```

**Revert `project.ts`** (drops `auditLog(...).catch(captureException)`) → the AC6 hardening test goes red:

```
× still resolves with the new key so the user is not locked out
```

With the bare `await auditLog(...)`, an audit-service failure rejects the mutation *after* the key was already rotated in the DB — the user is locked out (old key dead, new key never returned). That `.catch` is exactly what AC6 guards against.

**Reproduce**

```bash
mb=1f8a32a4a   # git merge-base HEAD origin/main
git checkout $mb -- langwatch/src/pages/settings/api-keys/ApiKeysSection.tsx \
                    langwatch/src/server/api/routers/project.ts
pnpm test:integration src/pages/settings/api-keys/__tests__/project-key-rotation.integration.test.tsx  # 🔴 5 failed
pnpm test:unit        src/server/api/routers/__tests__/project.regenerateApiKey.unit.test.ts            # 🔴 1 failed
git checkout HEAD -- langwatch/src/pages/settings/api-keys/ApiKeysSection.tsx \
                     langwatch/src/server/api/routers/project.ts                                        # restore → 🟢 7/7 each
```

**Green CI on HEAD `8a74f831b`** — `langwatch-app-ci` run [`27704558295`](https://github.com/langwatch/langwatch/actions/runs/27704558295): [`test-integration`](https://github.com/langwatch/langwatch/actions/runs/27704558295/job/81949649790) (1–6) · [`test-unit`](https://github.com/langwatch/langwatch/actions/runs/27704558295/job/81949649739) (1–4) · [`typecheck`](https://github.com/langwatch/langwatch/actions/runs/27704558295/job/81949649627) · [`lint`](https://github.com/langwatch/langwatch/actions/runs/27704558295/job/81949649776) · [`build`](https://github.com/langwatch/langwatch/actions/runs/27704558295/job/81949649613) · required [`langwatch-app-complete`](https://github.com/langwatch/langwatch/actions/runs/27704558295/job/81952358066) — all ✅.

*Browser-QA intentionally skipped (no app boot):* per the codified backend test-proof standard, the rotation surface is proven by the component test driving the actual Rotate control through the permission / success / failure paths — stronger than a screenshot, since it asserts behavior and the edge cases a screenshot can't show. The real-Postgres `project.regenerateApiKey.integration.test.ts` (AC2 round-trip) skips the testcontainers lane and is verified locally via `test:pg-integration`; it is not part of this red→green local run.

## Notes for review

- The server **DB** integration tests (`project.regenerateApiKey.integration.test.ts`) carry `describe.skipIf(isTestcontainersOnly)`, matching the pre-existing test in that file — they **skip** in the testcontainers CI lane and run against a real Postgres (verified locally via `test:pg-integration`). AC1/AC3-UI/AC4-UI/AC5/AC6 are also covered by the unit + component tests, which **do** run in CI.
- Two existing `ApiKeysSection`-rendering tests (`api-keys-ingestion-split`, `api-keys-scope-filter`) had their `~/utils/api` mock extended with the `project.regenerateApiKey` namespace, since the component now calls it at render.
- Spec: `specs/api-keys/project-key-rotation.feature` (behavior-level, alongside `unified-api-keys.feature`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added ability to rotate project API keys from Settings > API Keys page
  * New API key displayed after successful rotation
  * Permission-gated control for users with project management access
  * Confirmation dialog prevents accidental regeneration

* **Tests**
  * Added comprehensive integration and unit test coverage for key rotation workflows

### CODE REVIEW

**Comment by drewdrewthis** [comment]
## ✅ Live end-to-end demo — rotation works in the running app

Beyond the unit/component/integration tests (which mock the tRPC boundary), I drove a **real browser through the running app** to exercise the un-mocked `rendered page → tRPC → Postgres` wire that the tests can't.

**Decisive evidence (network):** `POST /api/trpc/project.regenerateApiKey?batch=1 → 200 OK` — the exact integration surface the tests mock.

**Flow verified (9 screenshots captured + inspected):**
1. Settings → API Keys shows the legacy **"Project API Key"** row with the new **"Rotate Project API Key"** control present (gated on `project:manage`) — key `sk-…wJ77`. _(AC1 control, AC3 UI gate)_
2. Clicking it opens the **"Regenerate API Key?"** confirm dialog with the "cannot be undone" warning.
3. Confirming → **"Token Created"** dialog shows the **new** key `sk-l…1tEW` once ("Copy this token now. You won't be able to see it again.") + a "Project API key rotated" toast. _(AC1: new key issued + shown once)_
4. The row refreshes to `sk-…1tEW` — the previous key is gone. _(AC2: rotation took effect / key changed)_

Combined with the server integration test (old key → no longer resolves via the real `TokenResolver`; new key → resolves scoped to the project; viewer → `UNAUTHORIZED`; audit row written) on a real Postgres, all 6 ACs are demonstrated end-to-end.

_Screenshots archived on the dev box at `/tmp/pr4894-demo/` (`01`…`09`). Demo used a throwaway local signup; no production data touched._

**Comment by drewdrewthis** [comment]
## Review verdict: READY

Multi-reviewer pass (principles, hygiene, security, test, proof) + personas on the full PR, then a scoped re-review of the review-thread fixes. The feature is correct and complete — proof-reviewer verified all 6 ACs map to real in-tree evidence, the live browser demo proved the un-mocked `page → tRPC → Postgres` wire, security review came back clean on every dimension, and CI is green.

All 6 review threads are resolved (3 fixed in `5481b9ae9`, 2 reasoned-resolve, 1 auto-resolved on push). No blocking change-requests.

### Review-thread fixes (commit `5481b9ae9`, re-reviewed)
- **Audit-log non-fatal** — `regenerateApiKey`'s `auditLog(...)` is now `.catch(captureException)` (the same pattern the `create` mutation uses) so an audit failure after the committed key-update can no longer return an error while withholding the rotated key. principles + security re-reviewed and confirmed: the `.catch` wraps only the audit promise, the update-failure path is unchanged, and fail-open-with-capture is the correct posture (the update commits before the audit, so the prior fatal behaviour gave no atomicity — only lockout). Backed by a regression unit test.
- **AC1 "shown once"** — added a component test asserting the rotated key is revealed in `TokenCreatedDialog` (stubs the lazy Shiki box to defeat dynamic-import flakiness). Dead duplicate mock + a confusing comment from that fix were removed in `8a74f831b`.
- **BDD restructure** + **unused-import** cleanups applied.

### Non-blocking findings (recommended follow-ups — not blockers)
1. **[security] Decide** — `captureException(auditError)` is bare; pass explicit `{ tags: { mutation: "regenerateApiKey" }, extra: { projectId } }` for actionable incident-response and to bound what Prisma error `meta` is forwarded. Posture is already sound (distinctId is `"server"`, no key material in the rejection); this is a hardening nicety.
2. **[test] New Issue** — AC2 / server-side AC3 / audit-row AC5 are covered by server integration tests that `skipIf(isTestcontainersOnly)` and so **skip in CI** (verified via the test-integration log `↓ 5 skipped`). Covered locally (real Postgres) + the browser demo; follows the existing repo pattern. Worth running the Postgres-only suites in CI repo-wide.
3. **[security] New Issue** — all four `onError` handlers in `ApiKeysSection` surface `error.message` in a toast; whole-file hardening pass.
4. **[hygiene] New Issue** — pre-existing `isAdmin = (orgMembers.data?.length ?? 0) > 0` gates edit/revoke by membership count rather than RBAC.

File: langwatch/src/pages/settings/api-keys/__tests__/project-key-rotation.integration.test.tsx [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+ * The unified-keys rework removed the UI to rotate the project base/legacy
+ * key. These tests cover restoring that control on the legacy "Project API
+ * Key" row: it is permission-gated on `project:manage`, opens the reused
+ * regenerate-confirm dialog, and drives the `project.regenerateApiKey`
+ * mutation. A failed rotation surfaces an error toast.
+ *
+ * We assert the mutation mock is called (and the error path toasts) without
+ * driving the success path through `TokenCreatedDialog` — that surface pulls
+ * in a dynamic ShikiCommandBox import that is flaky under jsdom.
+ */
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/pages/settings/api-keys/__tests__/project-key-rotation.integration.test.tsx]: Fixed in 5481b9ae9 — added a success-path test asserting the rotated key is revealed once in the Token Created dialog. The lazy Shiki box is mocked (next-dynamic stub) so the success render is no longer skipped for flakiness; the test drives onSuccess and asserts the new key appears.

File: langwatch/src/server/api/routers/__tests__/project.regenerateApiKey.integration.test.ts [hunk 2]
```diff
@@ -17,129 +18,258 @@ import { createInnerTRPCContext } from "../../trpc";
...
+
+    it("returns UNAUTHORIZED for nonexistent project (secure behavior - does not reveal project existence)", async () => {
+      await expect(
+        caller.project.regenerateApiKey({
+          projectId: "nonexistent-project-id",
+        }),
+      ).rejects.toMatchObject({
+        code: "UNAUTHORIZED",
+      });
     });
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/__tests__/project.regenerateApiKey.integration.test.ts]: Fixed in 5481b9ae9 — wrapped the two root-level it() into nested given/when describe blocks per the repo BDD standard.

File: langwatch/src/server/api/routers/__tests__/project.regenerateApiKey.unit.test.ts [hunk 3]
```diff
@@ -214,5 +214,28 @@ describe("project.regenerateApiKey mutation logic", () => {
...
+      );
+
+      await expect(
+        caller.regenerateApiKey({ projectId }),
+      ).rejects.toBeDefined();
+
+      expect(auditLog).not.toHaveBeenCalledWith(
+        expect.objectContaining({ action: "project.apiKey.regenerated" }),
+      );
+    });
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/__tests__/project.regenerateApiKey.unit.test.ts:238]: Fixed in 5481b9ae9 — the audit-log write is now non-fatal (catch + captureException, the same pattern the create mutation uses), so an audit failure after the key update can no longer surface an error while withholding the rotated key. Added a regression unit test asserting the mutation still resolves with the new key when the audit log throws.

File: langwatch/src/pages/settings/api-keys/ApiKeysSection.tsx [hunk 4]
```diff
@@ -312,9 +399,17 @@ export function ApiKeysSection({
...
+          All
+        </Badge>
+      );
     }
-    return <Badge size="sm" colorPalette="orange">Restricted</Badge>;
+    return (
+      <Badge size="sm" colorPalette="orange">
+        Restricted
+      </Badge>
+    );
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/pages/settings/api-keys/ApiKeysSection.tsx:412]: Out of scope for this PR. getPermissionBadge is pre-existing and not on the rotation path — the legacy Project API Key row hardcodes its All badge, so readonly labelling does not affect this feature. Tracked as a follow-up; resolving to keep the rotation change focused.

File: langwatch/src/pages/settings/api-keys/ApiKeysSection.tsx [hunk 5]
```diff
@@ -336,219 +431,264 @@ export function ApiKeysSection({
...
+                    <Table.ColumnHeader>Created</Table.ColumnHeader>
+                    <Table.ColumnHeader>Last Used</Table.ColumnHeader>
+                    <Table.ColumnHeader>Type</Table.ColumnHeader>
+                    <Table.ColumnHeader>Scope</Table.ColumnHeader>
+                    <Table.ColumnHeader>Permissions</Table.ColumnHeader>
+                    <Table.ColumnHeader width="100px"></Table.ColumnHeader>
+                  </Table.Row>
+                </Table.Header>
+                <Table.Body>
+                  {/* Project service key row — only shown when it survives the active scope filter */}
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/pages/settings/api-keys/ApiKeysSection.tsx:481]: Keeping. These are pre-existing comments not introduced by this change; the line-481 one documents the non-obvious scope-filter gating (why the legacy row may not render), which is a why and not a restatement. Out of scope for the rotation change; resolving.

**Review by rogeriochaves — APPROVED** [review]
